package/MDAnalysis/analysis/dssp/dssp.py

```diff
@@ -0,0 +1,364 @@
+    atoms : Union[Universe, AtomGroup]
+        input Universe or AtomGroup. In both cases, only protein residues will
+        be chosen. Heavy atoms of the protein are then selected by name
+        ("N", "CA", "C", "O O1 OT1"), and hydrogens are selected by name "H".
+    guess_hydrogens : bool, optional
+        whether you want to guess hydrogens positions, by default ``True``.
+        Guessing is made assuming perfect 120 degrees for all bonds that N
+        atom makes, and a N-H bond length of 1.01 A.
+        If ``guess_hydrogens`` is False, hydrogen atom positions on N atoms
+        will be parsed from the trajectory, except for the "hydrogen" atom
+        positions on PRO residues, and an N-terminal residue.
+
+    Raises
+    ------
+    ValueError
+        if ``guess_hydrogens`` is True but some non-PRO hydrogens are missing.
+
+    Examples
+    --------
+
+    For example, you can assign secondary structure for a single PDB file:
+
+    >>> from MDAnalysis.analysis.dssp import DSSP
+    >>> from MDAnalysisTests.datafiles import PDB
+    >>> import MDAnalysis as mda
+    >>> u = mda.Universe(PDB)
+    >>> run = DSSP(u).run()
+    >>> print("".join(run.results.dssp[0][:20]))
+    --EEEEE-----HHHHHHHH
+
+    Also, per-frame dssp assignment allows you to build average
+    secondary structure -- ``DSSP.results.dssp_ndarray`` holds
+    (n_frames, n_residues, 3) shape ndarray with one-hot encoding of
+    loop, helix and sheet, respectively:
+
+    >>> from MDAnalysis.analysis.dssp import translate, DSSP
+    >>> from MDAnalysisTests.datafiles import TPR, XTC
+    >>> u = mda.Universe(TPR, XTC)
+    >>> long_run = DSSP(u).run()
+    >>> mean_secondary_structure = translate(long_run.results.dssp_ndarray.mean(axis=0))
+    >>> print(''.join(mean_secondary_structure)[:20])
+    -EEEEEE------HHHHHHH
+
+    .. versionadded:: 2.8.0
+    """
+
+    def __init__(self, atoms: Union[Universe, AtomGroup], guess_hydrogens: bool = True):
+        self._guess_hydrogens = guess_hydrogens
+
+        ag: AtomGroup = (
+            atoms.select_atoms("protein") if isinstance(atoms, Universe) else atoms
+        )
+        super().__init__(ag.universe.trajectory)
+
+        # define necessary selections
+        # O1/OT1 is for C-terminal residue
+        heavyatom_names = ("N", "CA", "C", "O O1 OT1")
+        self._heavy_atoms: dict[str, "AtomGroup"] = {
+            t: ag.select_atoms(f"protein and name {t}") for t in heavyatom_names
+        }
+        self._hydrogens: list["AtomGroup"] = [
+            res.atoms.select_atoms("name H")
+            for res in ag.select_atoms("protein").residues
+        ]
+        # can't do it the other way because I need missing values to exist
+        # so that I could fill them in later
+
+        if not self._guess_hydrogens:
+            for calpha, hydrogen in zip(
+                self._heavy_atoms["CA"][1:], self._hydrogens[1:]
+            ):
+                if not hydrogen and calpha.resname != "PRO":
```

Comment:
I have added a `Warning` to the docstring -- hope that works fine.

Reply:
Thanks for adding the warning, however I'm still a bit confused with this https://github.com/MDAnalysis/mdanalysis/pull/4304/files#diff-cad8095a75b51cba3d6ef256c3c05073b457f309406ae4a94f45ac53257f3e81R202

It claims we're checking for "H" but we only expect one "H" according to the warning? (sorry I'm probably misreading something).

Reply:
Hi, I couldn't exactly follow your link -- for me it leads to a general page with all diffs for this PR. But anyway, we're selecting hydrogens with `name {hydrogen_name}`, which is supposed to select *the* hydrogen on the peptide bond's nitrogen, and we indeed expect only one (or zero, in case or prolines) of these hydrogens.

Reply:
The warning says that we expect exactly one H for non-PRO. Can you write the `if` with the following logic
```python
              if calpha.resname == "PRO":
                  # special case PRO
                  if hydrogen:
                       raise ValueError(f"PRO {calpha.resid} has a hydrogen {hydrogen}")
                  # else:
                  #     pass  # expected: PRO has no hydrogen
              elif not hydrogen or len(hydrogen) > 1:
                  # not a PRO and not exactly one HN: that's bad
                  raise ValueError(f"{calpha.resname} {calpha.resid} has not exactly one HN, instead it has {hydrogen}")
```

At least that's my understanding what we _should_ be seeing, @marinegor ?

(There's better code to be written than what I wrote but for following the logic I inserted a `pass` etc...)

Reply:
Oh, I see what you mean. Added a check for the number of hydrogens in the residue -- I hope that solves the issue.

Reply:
nice